Refactor extension organization

kgryte opened the pull request:
PR to address https://github.com/jupyterlab/jupyterlab-data-explorer/issues/83.

kgryte committed:
Refactor CSV viewer conversion

kgryte committed:
Update include filter to include subfolders

kgryte committed:
Fix typo

kgryte committed:
Add interface and add debugging hack

kgryte committed:
Remove extra space

kgryte committed:
Remove unused import

kgryte committed:
Add missing semicolon

kgryte committed:
Fix indentation

kgryte committed:
Add comment

kgryte committed:
Document type and create interface

kgryte committed:
Fix asterisk alignment

kgryte committed:
Mark interface definition as private

kgryte committed:
Mark internal class as private

kgryte committed:
Convert type definition to interface

kgryte committed:
Update comment

saulshanabrook commented on dataregistry-extension/src/csvviewer/activate.ts:
```diff
@@ -0,0 +1,129 @@
+/**
```
Instead of `activate.ts` maybe have `converters.ts` file that exports function from `registry` to array of converters.

kgryte replied:
My motivation: (1) 1 file, 1 concern and (2) testing.
​
In short, a guiding principle in Node.js development is to create files (modules) which export one thing (i.e., have a single concern), in order to encourage readability and testability.
​
If a file contains a single export, no need to track down a file's exports, which, if a file is long and not structured, can be tedious and time consuming.
​
And if a file has a single export, encourages writing tests which import that single export and directly test an API surface area. Often when everything is in long files, we see correspondingly long test files with correspondingly increased difficulty in achieving 100% code coverage. By having single file tests, easier to test using mocks via dependency injection (classic is proxyquire to proxy imported dependencies) and, as tests are documentation, easier to see how exported functionality should behave.
​
For example, take `activate.ts`. The file exports a single function which accepts two arguments and whose intent is to register "converters". When testing, we can import the sole export from this individual file, and, as Jupyter uses dependency injection, we can mock the second argument with an object containing a single method: `addConverter`, testing whether particular converters are registered, etc.
​
Similarly, I moved the converters to separate files based on the same idea. We should be able to test a particular converter, say, the one generated by export from `text_to_csv`, to ensure proper conversion behavior, such as the conversion of text "observable" to a CSV "observable", etc. If everything is in a single file and all the logic tightly coupled, this becomes rather difficult to test, especially thoroughly.
​
In a sense, each file having a single concern allows for, perhaps arguably, easier enforcement of API contracts.
​
Single-file-single-concern "style" can be found, to some degree, in Node.js core, and more so in Node.js modules written with the file system in mind and not, say, like jQuery, as a single file to be downloaded by a browser.
​
This "style" takes inspiration from the Unix philosophy in the sense that each module should do only one thing and do that thing well. Often, we think of this at the package level, but applies equally to the module/file layer.

saulshanabrook commented on dataregistry-extension/src/csvviewer/data_grid.ts:
```diff
@@ -0,0 +1,76 @@
+/**
```
Maybe call it CSVDataGrid to match the class? What do you think?

kgryte replied:
Can do, if think it is useful. The main idea here is that, to a module consumer, an implementation (and name) should be opaque and having to know an exported symbol identifier means having to know something about an implementation (namely, the identifiers in that file).
​
Here, my desire in the `csv_to_data_grid.ts` file is to convert from CSV to a data grid. The activation file does not need to know that the `data_grid.ts` file exports an extended class (i.e., a specialized data grid). To it, it can simply import from `data_grid.ts` a class for creating a data grid widget. So the naming of just `data_grid.ts` was to reflect that idea: from the `data_grid.ts` file, I can import a class for creating a data grid, assuming that whatever that class is can handle CSV data.
​
I updated the exported name in `data_grid.ts` to reflect this idea.

saulshanabrook commented on dataregistry-extension/src/csvviewer/data_grid.ts:
```diff
@@ -0,0 +1,76 @@
+     * @private
+     * @param data - CSV data
+     */
+    function onData(data: string) {
```
Maybe make this a method?

kgryte replied:
Yes, this could make sense. Generally, I avoid hanging private (implementation) methods off prototypes to ensure encapsulation, unless their are clear performance advantages to do otherwise.
​
Which raises the question: how frequently is the `before-attach` message sent? If often, then I should definitely create an private internal method; but, if not, then probably fine to leave as is.

saulshanabrook commented on dataregistry-extension/src/csvviewer/data_grid.ts:
```diff
@@ -0,0 +1,76 @@
+ * Exports.
+ */
+export { CSVDataGrid };
+export default CSVDataGrid;
```
I don't think. we need a default export

kgryte replied:
In general, my preference would be that every file have **only** a default export and no named exports for reasons mentioned earlier. However, I mainly included named exports because, with few exceptions, named exports are the norm in this project.
​
This has been debated somewhat in the Node.js community where some developers do both in order to allow users to pick whichever style they prefer and whichever style matches their project conventions.

saulshanabrook commented on dataregistry-extension/src/mimetypes.ts:
```diff
@@ -0,0 +1,28 @@
+/**
```
I don't think we need a mimetypes file

kgryte replied:
The reason I did was because I noticed MIME type magic constants which were repeated in various places (e.g., `if type === "text/csv" then...`).
​
Whenever I see repeated magic constants, I sense we'd be better off defining these in a single place (file) which we can then import.
​
Otherwise, we run the risk of typos and subtle bugs. For example, if I forget the `i` in `"text/plain"` when doing a comparison or when providing as an argument, TypeScript won't be able to catch this as a type error. I'd still be comparing strings or providing a string argument.
​
In this case, I'd only be able to catch the typo during runtime tests, not at compile time.
​
However, if I import a MIME types database and misspell a property name (most likely a name which does not exist), then I am more likely to get a compile error because I'd provide, e.g., `undefined` as an argument and not a string, and avoid the runtime bug.

saulshanabrook commented on dataregistry-extension/src/datatypes.ts:
```diff
@@ -0,0 +1,78 @@
+/**
+ * Exports.
+ */
+export { datatypes };
```
I think this should just export each data type, like:

```
export const CSVDataType = new DataTypeNoArgs<Observable<string>>("text/csv");
```

We don't need to copy all these other ones. They are defined in other files.

kgryte replied:
There may be some confusion here. This file is not defined in the `csvviewer` folder by in the parent directory.
​
In short, the idea is that, rather than having each data converter extension export various data types, thus requiring us to read individual files to see what data types are available, we should provide a single "database" of datatypes.
​
This allows us, when creating converters, to only need to consult one file, rather than many, in order to know what datatypes are implemented in the dataregistry.
​
If we have datatypes scattered across different files, we'll have to engage in an expedition to find files exporting data types and then, even supposing we codify a file in which these are defined, we have to read that separate files to know the identifier name which is exported (e.g., `CSVDataType` or `DatasetsFileDataType`, etc).
​
Instead, I'd argue that we'd be better off centralizing the declaration of data types in a single file (i.e., database). Meaning, in this refactoring, I will move the data types declarations to this `datatypes.ts` file. Then, whenever I'd want to implement a new converter, I'd only need to consult a single file.

saulshanabrook commented on dataregistry-extension/src/csvviewer/index.ts:
```diff
@@ -0,0 +1,25 @@
+import { JupyterFrontEndPlugin } from "@jupyterlab/application";
+import { IRegistry } from "./../registry";
+import { activate } from "./activate";
+
```
I think this should also export all from the datatypes file, so that other modules can refer to this datatypes

kgryte committed:
Update class description

kgryte committed:
Rename export

kgryte committed:
Move converters to individual files

kgryte committed:
Use default export convention

kgryte committed:
Remove MIME type database and use data type database as source of truth

kgryte committed:
Cache function result

kgryte committed:
Move implementations to separate packages

kgryte committed:
Add npm config file

kgryte committed:
Move packages to `packages` folder

kgryte committed:
Update paths

kgryte committed:
Add root TypeScript config

kgryte committed:
Add root ESLint config file

kgryte committed:
Update script commands and remove ESLint config files

kgryte committed:
Update dev deps

kgryte committed:
Add Lerna configuration file

kgryte committed:
Add meta package

kgryte committed:
Update scripts

kgryte committed:
Update scripts

kgryte committed:
Update settings

kgryte committed:
Reformat to ensure consistent with other config files

kgryte committed:
Update configuration

kgryte committed:
Update scripts

kgryte committed:
Set npm arguments

kgryte committed:
Update TypeScript configuration files and reformat relative paths

kgryte committed:
Remove meta package

kgryte committed:
Disable type checking of all declaration files

kgryte committed:
Remove unused import

kgryte committed:
Use `any` as type return value to allow build

kgryte committed:
Disable build after linking

kgryte committed:
Disable building after unlink

kgryte committed:
Fix script for automatic recompilation

kgryte committed:
Allow dev deps to opt in to patch and minor upgrades

kgryte committed:
Allow deps to opt in to patch and minor upgrades

kgryte committed:
Disable TypeScript compiler errors for offending line

kgryte committed:
Switch to yarn

kgryte committed:
Update scripts to use `jlpm`

kgryte committed:
Reorder fields

kgryte committed:
Remove relative path prefix due to Jupyter build path resolution bug

kgryte committed:
Remove exports for packages which are independently registered

kgryte committed:
Merge branch 'master' of https://github.com/jupyterlab/jupyterlab-data-explorer into extension-folders

kgryte committed:
Add dev dep to create a local registry proxy

kgryte committed:
Remove unused variable

kgryte committed:
Fix typo

kgryte committed:
Fix duplicate identifier and obsolete import

kgryte committed:
Add scripts

kgryte committed:
Add scripts

kgryte committed:
Simplify registry initialization script

kgryte committed:
Add scripts

kgryte committed:
Add development guide

kgryte committed:
Add organization and move core concepts

kgryte committed:
Add note

kgryte committed:
Add link

kgryte committed:
Add links

kgryte committed:
Update copy

kgryte committed:
Update copy

kgryte committed:
Update copy

kgryte committed:
Fix missing word

kgryte committed:
Update copy

kgryte committed:
Update copy

kgryte committed:
Update copy

kgryte committed:
Add word

kgryte committed:
Update copy

kgryte committed:
Update copy

kgryte committed:
Fix broken links

kgryte committed:
Normalize list style

kgryte committed:
Add EditorConfig

kgryte committed:
Add link to EditorConfig

kgryte committed:
Update copy and links

kgryte committed:
Fix link

kgryte committed:
Move assignment

kgryte committed:
Fix missing period

kgryte commented:
@saulshanabrook This PR still needs some small changes, but the broad strokes are there. Obviously, many files have changed. I wrote up a development guide and setup a number of `npm` scripts.

I think it would be beneficial to get your feedback, if any. You may want to pull this down separately, follow the development guide, and see if you can run everything locally.

My sense is that, once I add a few touch-ups, would be good to merge this as is, so that we can begin working from the reorganized structure. I can move the remaining converters to individual extension packages in a piecemeal fashion.

kgryte committed:
Merge branch 'master' of https://github.com/jupyterlab/jupyterlab-data-explorer into extension-folders

kgryte committed:
Update lint scripts

saulshanabrook committed:
Revert "Active URL fixes"

kgryte committed:
Resolve lint errors

saulshanabrook commented:
>  I  think it would be beneficial to get your feedback, if any. You may want  to pull this down separately, follow the development guide, and see if  you can run everything locally. My sense is that, once I add a few touch-ups, would be good to merge  this as is, so that we can begin working from the reorganized structure.  I can move the remaining converters to individual extension packages in  a piecemeal fashion.


Sounds good, I will give this a go.

saulshanabrook commented on package.json:
```diff
@@ -1,16 +1,46 @@
+    "uninstall:extensions": "jupyter labextension uninstall --all --no-build",
+    "unlink:packages": "lerna run unlink"
+  },
+  "devDependencies": {
```
Do you have a sense of what dev dependencies should be in the root package vs the child packages? It seems like some are added in both places.

kgryte replied:
Yeah, I can remove `rimraf` and `typescript` from the individual packages, I believe.

saulshanabrook commented:
So far I have been using prettier on this codebase to maintain a consistent style. Can I go ahead and reformat all of your work with it?

saulshanabrook committed:
Merge origin/master into pr/kgryte/93

kgryte commented:
I did not see a build command for reformatting, so I had yet to do so.

saulshanabrook commented on docs/development.md:
```diff
@@ -0,0 +1,264 @@
+<!-- 
```
Thoughts on moving this to `CONTRIBUTING.md` so [github picks it up](https://help.github.com/en/articles/setting-guidelines-for-repository-contributors#adding-a-contributing-file)?

kgryte replied:
I typically view the `CONTRIBUTING.md` file to be a bit higher level, linking to code of conducts, including a developer's certificate of origin, and linking to other docs necessary for contributing, such as a developer environment guide, style guides, etc. For an example, see [here](https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md). On reflection, some overlap between them. I am fine with consolidating if you think it best.

saulshanabrook replied:
I am fine with merging this as is and iterating. I am fine with either.

kgryte replied:
Added a contributing doc. In the abstract, I see "contributing" and "development" as two separate, but possible overlapping things. For example, I may want to locally develop, but never contribute, and I may want to contribute, but never develop. The contributing guide I included is based on the one I linked to above and includes a bit more hand holding for neophytes. Maybe this is unnecessary.

saulshanabrook replied:
That makes sense to me! Thank you.

kgryte committed:
Remove duplicate dependencies

saulshanabrook commented:
> I did not see a build command for reformatting, so I had yet to do so.

Yeah there isn't one. We should add one, at least for linting to verify everything is prettified. I have my editor set up to reformat on save.

kgryte commented:
I'd say that, ideally, we'd make such development processes editor agnostic. :)

saulshanabrook commented:
> I'd say that, ideally, we'd make such development processes editor agnostic. :)

Yep that sounds ideal :)

kgryte committed:
Ignore lockfiles

kgryte committed:
Move images

saulshanabrook commented on docs/development.md:
````diff
@@ -0,0 +1,263 @@
+While not required, you are encouraged to create an [Anaconda][anaconda] environment.
+
+```bash
+$ conda create -n jupyterlab-data-explorer -c conda-forge python=3.6 -c anaconda jupyterlab nodejs
````
```suggestion
$ conda create -n jupyterlab-data-explorer -c conda-forge python=3.6 jupyterlab nodejs
```

kgryte replied:
Okay. Question. I essentially took this from the Jupyter [docs](https://jupyterlab.readthedocs.io/en/stable/developer/extension_tutorial.html#install-nodejs-jupyterlab-etc-in-a-conda-environment). The `-c anaconda` is specifying that the remaining arguments be installed via the  Anaconda "channel", correct? If this is unnecessary, why do the Jupyter docs show otherwise (or maybe I am misreading them)?

saulshanabrook replied:
I believe you have to put all the channel names first. In the jupyter docs, it does this:

```
conda create -n jupyterlab-ext --override-channels --strict-channel-priority -c conda-forge -c anaconda jupyterlab cookiecutter nodejs git
```

I always prefer to just to use conda-forge and not the anaconda channel. I know we had a long discussion about this over slack at quansight just this week... @scopatz Do you have any recommendations on including anaconda channel or just conda forge in this command?

kgryte replied:
Updated.

scopatz replied:
yeah, I recommend conda-forge

kgryte committed:
Add contributing guide

kgryte committed:
Fix language

kgryte committed:
Add backticks

kgryte committed:
Fix command

saulshanabrook commented:
What do you think about having the packages that are extensions end in `-extension`? That's been our custom in JupyterLab core. So `dataregistry-csvviewer` -> `dataregistry-csvviewer-extension` and `dataregistry-registry` -> `dataregistry-registry-extension`?

saulshanabrook commented on package.json:
```diff
@@ -1,16 +1,46 @@
+    "lint:typescript": "eslint packages/**/src/*.ts",
+    "rebuild:packages": "jlpm run clean:packages && jlpm run build:packages",
+    "registry": "verdaccio",
+    "registry:init": "npm adduser --registry http://localhost:4873 && lerna exec --scope dataregistry-csvviewer --scope dataregistry-registry -- npm publish --registry http://localhost:4873",
```
I had to change this line to specify the package names instead of the paths:

```suggestion
    "registry:init": "npm adduser --registry http://localhost:4873 && lerna exec --scope @jupyterlab/dataregistry-csvviewer --scope @jupyterlab/dataregistry-registry -- npm publish --registry http://localhost:4873",
```

Otherwise I would get:

```
jlpm run registry:init
yarn run v1.15.2
$ npm adduser --registry http://localhost:4873 && lerna exec --scope dataregistry-csvviewer --scope dataregistry-registry -- npm publish --registry http://localhost:4873
Username: abc
Password:
Email: (this IS public) abc@abc.com
Logged in as abc on http://localhost:4873/.
lerna notice cli v3.16.4
lerna info versioning independent
lerna info filter [ 'dataregistry-csvviewer', 'dataregistry-registry' ]
lerna ERR! EFILTER No packages remain after filtering [ 'dataregistry-csvviewer', 'dataregistry-registry' ]
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

kgryte replied:
Ah, you're right! I misunderstood the scope filters. Will fix now.

kgryte replied:
Pushed a patch.

kgryte commented:
I thought about that, but then disliked how long the package names were getting. However, I recognize that this seems to be a growing convention within Jupyter and the ecosystem. I'll go with whatever you think best on this, as you have a better pulse on where things are going and the attendant conventions.

kgryte committed:
Fix filter flags

saulshanabrook commented:
> I thought about that, but then disliked how long the package names were getting.

I think it's useful, just as another signal to developers which things expose an extension and which do not.

kgryte commented:
Rog. Will rename shortly.

saulshanabrook committed:
Split registration into two commands

I found this helpful when having to re-try the publish command
without having to add a new user

saulshanabrook committed:
Update dataregistry version

While this PR was in motion, we bumped the version of the dataregistry
so the two extensions should also be bumped

saulshanabrook committed:
add dependency on dataregistry from csv viewer

kgryte commented on packages/dataregistry-csvviewer/package.json:
```diff
@@ -37,6 +37,7 @@
     "@jupyterlab/application": "^1.0.0",
     "@jupyterlab/csvviewer": "^1.0.0",
     "@jupyterlab/dataregistry": "^3.0.0",
+    "@jupyterlab/dataregistry-registry": "^0.0.0",
```
See L44. This is already included.

saulshanabrook replied:
Thanks!

kgryte committed:
Rename packages

saulshanabrook committed:
Revert "add dependency on dataregistry from csv viewer"

df8bc5a5842d54dca0c6dd5ab6c856f4bdc3b477

kgryte committed:
Merge branch 'extension-folders' of https://github.com/kgryte/jupyterlab-data-explorer into extension-folders

kgryte committed:
Merge branch 'extension-folders' of https://github.com/kgryte/jupyterlab-data-explorer into extension-folders

saulshanabrook commented on package.json:
```diff
@@ -1,16 +1,48 @@
+    "all": "jlpm run clean && jlpm run build && jlpm run build:jupyter",
+    "all:watch": "jlpm run clean && jlpm run build && jlpm run build:jupyter:watch",
+    "build": "jlpm run build:dev",
+    "build:dev": "jlpm install && jlpm run clean:packages && lerna link && jlpm run build:packages && jlpm run link:packages",
```
Any reason  to use `lerna link` over `lerna bootstrap`? I don't know the difference but the lerna readme just mentions bootstrap.

kgryte replied:
Yeah, did not want Lerna executing any of the other commands in `bootstrap` beside linking.

kgryte replied:
See [here](https://github.com/lerna/lerna/tree/master/commands/bootstrap#usage).

saulshanabrook replied:
Ah thank you! Sounds good

saulshanabrook commented:
Oh no, I am getting the dreaded error!

> error Couldn't find package "@jupyterlab/dataregistry-registry@0.0.0" required by "@jupyterlab/dataregistry-csvviewer@file:../extensions/jupyterlab-dataregistry-csvviewer-0.0.0-718c020eb361908eddb162f4fc6b1ccd253c90ba.tgz" on the "npm" registry.

I think the registry isn't being passed in through the env variable for me, since it isn't hitting the API

kgryte commented:
And now I am encountering the same error. **sigh**

kgryte committed:
Fix package names

kgryte committed:
Fix missing dep

kgryte committed:
Add local registry

kgryte committed:
Add and update scripts

kgryte committed:
Add comment

kgryte committed:
Remove script

kgryte committed:
Update script

saulshanabrook commented on package.json:
```diff
@@ -1,16 +1,50 @@
+    "build": "jlpm run build:dev",
+    "build:dev": "jlpm install && jlpm run clean:packages && lerna link && jlpm run build:packages && jlpm run link:packages && jupyter labextension list",
+    "build:jupyter": "jupyter lab build || (jlpm run build:jupyter:remediate && jupyter lab build)",
+    "build:jupyter:remediate": "curr_dir=$PWD && (cd $(jupyter lab path | head -n1 | sed s/Application\\ directory:\\ *//)/staging && echo $PWD && echo '\"@jupyterlab:registry\" \"http://localhost:4873\"' >> .yarnrc && cat .yarnrc && YARN_REGISTRY=http://localhost:4873 node ./yarn.js install --non-interactive && cd $curr_dir) || cd $curr_dir",
```
Nothing will go wrong here! :)

kgryte replied:
Nope. Nothing could go wrong. It would be better if we only performed the "remediation" step when we encounter the npm registry error, but that would require parsing the error log, which seems a bit heavy handed atm. Assuming we encounter a non-registry error which is reproducible, the second `jupyter lab build` should simply generate the same error. Or at least that is the hope...

saulshanabrook commented on packages/dataregistry/src/cachedObservable.ts:
```diff
@@ -139,7 +139,7 @@ export class CachedObservable<T> extends Observable<T> {
           subscriber.next(state.value);
           return () => {};
         default:
-          const _exhaustiveCheck: never = state;
```
Nice! This is great, since it also preserves verifying at compile time that  the check is exhaustive. Thanks.

saulshanabrook commented:
Unfortunately, I am still getting the error. I think the issue is that after you add the yarnrc we run jupyter lab build, which wipes it out.

saulshanabrook commented:
This has been coming up for others as well, I am going to bring it up at tomorrows dev meeting: https://github.com/jupyterlab/jupyterlab#weekly-dev-meeting

kgryte committed:
Fix command

saulshanabrook committed:
Remove local registry config so I dont get an error on initial install

Otherwise I get this error when installing for the first time

```
conda activate jupyterlab-data-explorer
jlpm install
yarn install v1.15.2
[1/4] 🔍  Resolving packages...
error Couldn't find package "@jupyterlab/application@^1.0.0" required by "@jupyterlab/dataregistry-csvviewer-extension@0.0.0" on the "npm" registry.
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
warning workspace-aggregator-4012d5a9-bf71-42b6-9a7f-f1eae0bda212 > @jupyterlab/dataregistry > rxjs-spy > circular-json@0.5.9: CircularJSON is in maintenance only, flatted is its successor.
warning workspace-aggregator-4012d5a9-bf71-42b6-9a7f-f1eae0bda212 > @jupyterlab/dataregistry-extension > @nteract/data-explorer > semiotic > prop-types > fbjs > core-js@1.2.7: core-js@<2.6.8 is no longer maintained. Please, upgrade to core-js@3 or at least to actual version of core-js@2.
warning workspace-aggregator-4012d5a9-bf71-42b6-9a7f-f1eae0bda212 > @jupyterlab/dataregistry > jest > jest-cli > jest-config > jest-environment-jsdom > jsdom > left-pad@1.3.0: use String.prototype.padStart()
Error: Couldn't find package "@jupyterlab/application@^1.0.0" required by "@jupyterlab/dataregistry-extension@3.0.0" on the "npm" registry.
    at /usr/local/miniconda3/envs/jupyterlab-data-explorer/lib/python3.6/site-packages/jupyterlab/staging/yarn.js:37186:17
    at Generator.throw (<anonymous>)
    at step (/usr/local/miniconda3/envs/jupyterlab-data-explorer/lib/python3.6/site-packages/jupyterlab/staging/yarn.js:310:30)
    at /usr/local/miniconda3/envs/jupyterlab-data-explorer/lib/python3.6/site-packages/jupyterlab/staging/yarn.js:323:13
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
Error: Couldn't find package "@jupyterlab/application@^1.0.0" required by "@jupyterlab/dataregistry-registry-extension@0.0.0" on the "npm" registry.
    at /usr/local/miniconda3/envs/jupyterlab-data-explorer/lib/python3.6/site-packages/jupyterlab/staging/yarn.js:37186:17
    at Generator.throw (<anonymous>)
    at step (/usr/local/miniconda3/envs/jupyterlab-data-explorer/lib/python3.6/site-packages/jupyterlab/staging/yarn.js:310:30)
    at /usr/local/miniconda3/envs/jupyterlab-data-explorer/lib/python3.6/site-packages/jupyterlab/staging/yarn.js:323:13
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
Error: Couldn't find package "@jupyterlab/csvviewer@^1.0.0" required by "@jupyterlab/dataregistry-csvviewer-extension@0.0.0" on the "npm" registry.
    at /usr/local/miniconda3/envs/jupyterlab-data-explorer/lib/python3.6/site-packages/jupyterlab/staging/yarn.js:37186:17
    at Generator.throw (<anonymous>)
    at step (/usr/local/miniconda3/envs/jupyterlab-data-explorer/lib/python3.6/site-packages/jupyterlab/staging/yarn.js:310:30)
    at /usr/local/miniconda3/envs/jupyterlab-data-explorer/lib/python3.6/site-packages/jupyterlab/staging/yarn.js:323:13
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
Error: Couldn't find package "@jupyterlab/csvviewer@^1.0.0" required by "@jupyterlab/dataregistry-extension@3.0.0" on the "npm" registry.
    at /usr/local/miniconda3/envs/jupyterlab-data-explorer/lib/python3.6/site-packages/jupyterlab/staging/yarn.js:37186:17
    at Generator.throw (<anonymous>)
    at step (/usr/local/miniconda3/envs/jupyterlab-data-explorer/lib/python3.6/site-packages/jupyterlab/staging/yarn.js:310:30)
    at /usr/local/miniconda3/envs/jupyterlab-data-explorer/lib/python3.6/site-packages/jupyterlab/staging/yarn.js:323:13
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
Error: Couldn't find package "@jupyterlab/apputils@^1.0.0" required by "@jupyterlab/dataregistry-extension@3.0.0" on the "npm" registry.
    at /usr/local/miniconda3/envs/jupyterlab-data-explorer/lib/python3.6/site-packages/jupyterlab/staging/yarn.js:37186:17
    at Generator.throw (<anonymous>)
    at step (/usr/local/miniconda3/envs/jupyterlab-data-explorer/lib/python3.6/site-packages/jupyterlab/staging/yarn.js:310:30)
    at /usr/local/miniconda3/envs/jupyterlab-data-explorer/lib/python3.6/site-packages/jupyterlab/staging/yarn.js:323:13
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
Error: Couldn't find package "@jupyterlab/filebrowser@^1.0.0" required by "@jupyterlab/dataregistry-extension@3.0.0" on the "npm" registry.
    at /usr/local/miniconda3/envs/jupyterlab-data-explorer/lib/python3.6/site-packages/jupyterlab/staging/yarn.js:37186:17
    at Generator.throw (<anonymous>)
    at step (/usr/local/miniconda3/envs/jupyterlab-data-explorer/lib/python3.6/site-packages/jupyterlab/staging/yarn.js:310:30)
    at /usr/local/miniconda3/envs/jupyterlab-data-explorer/lib/python3.6/site-packages/jupyterlab/staging/yarn.js:323:13
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
Error: Couldn't find package "@jupyterlab/notebook@^1.0.0" required by "@jupyterlab/dataregistry-extension@3.0.0" on the "npm" registry.
    at /usr/local/miniconda3/envs/jupyterlab-data-explorer/lib/python3.6/site-packages/jupyterlab/staging/yarn.js:37186:17
    at Generator.throw (<anonymous>)
    at step (/usr/local/miniconda3/envs/jupyterlab-data-explorer/lib/python3.6/site-packages/jupyterlab/staging/yarn.js:310:30)
    at /usr/local/miniconda3/envs/jupyterlab-data-explorer/lib/python3.6/site-packages/jupyterlab/staging/yarn.js:323:13
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
Error: Couldn't find package "@jupyterlab/coreutils@^3.0.0" required by "@jupyterlab/dataregistry-extension@3.0.0" on the "npm" registry.
    at /usr/local/miniconda3/envs/jupyterlab-data-explorer/lib/python3.6/site-packages/jupyterlab/staging/yarn.js:37186:17
    at Generator.throw (<anonymous>)
    at step (/usr/local/miniconda3/envs/jupyterlab-data-explorer/lib/python3.6/site-packages/jupyterlab/staging/yarn.js:310:30)
    at /usr/local/miniconda3/envs/jupyterlab-data-explorer/lib/python3.6/site-packages/jupyterlab/staging/yarn.js:323:13
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
Error: Couldn't find package "@jupyterlab/coreutils@^3.0.0" required by "@jupyterlab/dataregistry-registry-extension@0.0.0" on the "npm" registry.
    at /usr/local/miniconda3/envs/jupyterlab-data-explorer/lib/python3.6/site-packages/jupyterlab/staging/yarn.js:37186:17
    at Generator.throw (<anonymous>)
    at step (/usr/local/miniconda3/envs/jupyterlab-data-explorer/lib/python3.6/site-packages/jupyterlab/staging/yarn.js:310:30)
    at /usr/local/miniconda3/envs/jupyterlab-data-explorer/lib/python3.6/site-packages/jupyterlab/staging/yarn.js:323:13
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
Error: Couldn't find package "@jupyterlab/docmanager@^1.0.0" required by "@jupyterlab/dataregistry-extension@3.0.0" on the "npm" registry.
    at /usr/local/miniconda3/envs/jupyterlab-data-explorer/lib/python3.6/site-packages/jupyterlab/staging/yarn.js:37186:17
    at Generator.throw (<anonymous>)
    at step (/usr/local/miniconda3/envs/jupyterlab-data-explorer/lib/python3.6/site-packages/jupyterlab/staging/yarn.js:310:30)
    at /usr/local/miniconda3/envs/jupyterlab-data-explorer/lib/python3.6/site-packages/jupyterlab/staging/yarn.js:323:13
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
Error: Couldn't find package "@jupyterlab/docregistry@^1.0.4" required by "@jupyterlab/dataregistry-extension@3.0.0" on the "npm" registry.
    at /usr/local/miniconda3/envs/jupyterlab-data-explorer/lib/python3.6/site-packages/jupyterlab/staging/yarn.js:37186:17
    at Generator.throw (<anonymous>)
    at step (/usr/local/miniconda3/envs/jupyterlab-data-explorer/lib/python3.6/site-packages/jupyterlab/staging/yarn.js:310:30)
    at /usr/local/miniconda3/envs/jupyterlab-data-explorer/lib/python3.6/site-packages/jupyterlab/staging/yarn.js:323:13
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
Error: Couldn't find package "@jupyterlab/rendermime@^1.0.0" required by "@jupyterlab/dataregistry-extension@3.0.0" on the "npm" registry.
    at /usr/local/miniconda3/envs/jupyterlab-data-explorer/lib/python3.6/site-packages/jupyterlab/staging/yarn.js:37186:17
    at Generator.throw (<anonymous>)
    at step (/usr/local/miniconda3/envs/jupyterlab-data-explorer/lib/python3.6/site-packages/jupyterlab/staging/yarn.js:310:30)
    at /usr/local/miniconda3/envs/jupyterlab-data-explorer/lib/python3.6/site-packages/jupyterlab/staging/yarn.js:323:13
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
```

saulshanabrook commented on packages/dataregistry-csvviewer-extension/src/datatypes.ts:
```diff
@@ -0,0 +1,32 @@
+import { Widget } from "@phosphor/widgets";
+import { Observable } from "rxjs";
+
+// TODO: refactor
```
This looks good to me at the moment. The datatypes will be properly typed, so if you do `datatypes.csv` it will know the type. What do you have in mind for refactor?

kgryte replied:
I have the TODO here as this is a temporary hack until each of the converters is a separate extension. E.g., instead of L26, we'd import the data type from the widget data converter extension.

saulshanabrook commented on packages/dataregistry-csvviewer-extension/src/csv_to_data_grid.ts:
```diff
@@ -0,0 +1,67 @@
+    "to": datatypes.widget
+  };
+
+  return createConverter(conversion, convert);
```
I prefer the functions inline here, since the typescript know the types of everything. Otherwise, you have to manually specify the types. Like below, `convert` has no return annotation, so it is not verified that it is correct. It's a pain to add those annotations so if we have the functions inline, then we dont have to.

kgryte replied:
Yeah, I ended up frustrated with inlining because you cannot readily document inlined functions and cannot elevate them to a higher scope.

I also look at this another way:

> what is the minimum contract necessary for the `convert` function to work and for someone (e.g., an outside developer reading the source code for this extension) to be able to grok ?

Based on how `createConverter` works, I can already assume that `convert` is going to be provided an object containing data of the correct "type", that is part of the `createConverter` "contract", and we've already defined in the `conversion` object the data type in the return object. 

In short, TMK, we don't lose any "correctness" by using general types for what is an internal function interface.

And personally, and this is one of my gripes about using TypeScript, I shouldn't have to choose between precise and exhaustive types and no documentation/scope hoisting at all. TypeScript effectively encourages devs to inline functions (often via arrow functions in order to avoid explicit typing) which is not, IMO, good practice for a variety of reasons (perf, docs, debugging, readability, etc).

This stated, I can modify to the conventions of this project, if this is a blocker for merge. :)

saulshanabrook replied:
Thanks for the explanation here, that makes sense! I haven't thought as much about the downsides of Typescript's strictness, I just let it tell me what to do :)

Let's merge this in.

saulshanabrook commented on packages/dataregistry-csvviewer-extension/src/text_to_csv.ts:
```diff
@@ -0,0 +1,65 @@
+ *
+ * @private
+ */
+interface Data {
```
Again here, since the functions are not inline you have to manually annotate them with types, and this type is too general.

saulshanabrook commented:
At long last, the build succeeds! I followed the dev instructions and it worked. I also tested changing some code using the watch commands and that worked as well. I just left a few more comments, about some type annotations.

TLDR: Keeping callbacks inline means that you don't have to define type annotations for them, which keeps code concise and typed strongly.

kgryte committed:
Move functions to parent scope
